cEP-0018.md

```diff
@@ -272,3 +272,157 @@ class TestBear(ASTBear):
             """
         ...
 ```
+
+## Providing Fixes with the library
+
+ANTLR has a `HiddenTokenStream` which recieves all the tokens irrelevant to the
+parser such as whitespaces. These are not passed on to the parser and as a
+result are not returned when simply fetching text from a node in parse tree.
+
+Consider the code:
+
+```python
+def do_nothing():
+    pass
+
+def run_coala(console_printer=None,
+              log_printer=None,
+              print_results=do_nothing,
+              print_section_beginning=do_nothing,
+              nothing_done=do_nothing,
+              autoapply=True,
+              force_show_patch=False,
+              arg_parser=None,
+              arg_list=None,
+              args=None,
+              debug=False):
+    """
+    This function signature has been taken from coala-core for fairly complex
+    example.
+    """
+    print('This function will do something')
+    if (console_printer and
+        log_printer and
+        not (args == None and debug == False)
+        ):
+       print('These conditions were made to trick !')
+       print(' !(a & b) == !a or !b')
+```
+
+The corresponding parse-tree:
+![example_tree_complex](https://user-images.githubusercontent.com/10978108/43359512-e38dbfa0-92c0-11e8-8f5b-fab244d27ec0.png)
+
+After converting this source back using:
+
+```python
+#!/usr/bin/env python
+from antlr4 import *
+from Python3Lexer import Python3Lexer
+from Python3Parser import Python3Parser
+from Python3Visitor import Python3Visitor
+
+import sys
+
+class Visitor(Python3Visitor):
+    def visitFile_input(self, ctx):
+        print(ctx.getText())
+
+def main(args):
+    file = FileStream(args[1])
+    lexer = Python3Lexer(file)
+    stream = CommonTokenStream(lexer)
+    parser = Python3Parser(stream)
+
+    tre = parser.file_input()
+    if parser.getNumberOfSyntaxErrors()!=0:
+        print("File contains {} "
+              "syntax errors".format(parser.getNumberOfSyntaxErrors()))
+        return
+
+    visitor = Visitor()
+    visitor.stream = stream
+    visitor.visit(tre)
+
+if __name__ == '__main__':
+    main(sys.argv)
+```
+
+We get the following output:
+
+```python
+defdo_nothing():
+    pass
+
+defrun_coala(console_printer=None,log_printer=None,print_results=do_nothing,print_section_beginning=do_nothing,nothing_done=do_nothing,autoapply=True,force_show_patch=False,arg_parser=None,arg_list=None,args=None,debug=False):
+    """
+    This function signature has been taken from coala-core for fairly complex
+    example.
+    """
+print('This function will do something')
+if(console_printerandlog_printerandnot(args==Noneanddebug==False)):
+       print('These conditions were made to trick !')
+print(' !(a & b) == !a or !b')
+
+
+<EOF>
+```
+
+The most important points to note in the above are (Especially with respect
+to a `Python` grammar scenario):
+
+- All spaces within two tokens has been removed unless present inside a string
```

Comment:
your converter should add whitespace where it is legitimate to do that. (even this may be hard to do 'correctly', but it should be easy to do 80% of the time)

maybe have two printouts -- one raw, and the other with whitespace added.
Then you still need to show/explain that the whitespace added isnt the same whitespace as the original.